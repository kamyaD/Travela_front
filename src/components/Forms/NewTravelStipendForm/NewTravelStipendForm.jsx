import React, { PureComponent } from 'react';
import { PropTypes } from 'prop-types';
import { FormContext, getDefaultBlanksValidatorFor } from '../FormsAPI';
import TravelStipendFieldset from './FormFieldsets/TravelStipendDetails';
import SubmitArea from './FormFieldsets/SubmitArea';
import './NewTravelStipendForm.scss';

class NewTravelStipendForm extends PureComponent {
  constructor(props) {
    super(props);
    this.defaultState = {
      values: {
        center: '',
        stipend: '',
      },
      errors: {},
      hasBlankFields: true,
      isValidAmount: true
    };
    this.state = { ...this.defaultState };
    this.validate = getDefaultBlanksValidatorFor(this);
  }

  componentDidMount() {
    this.updateStateData(this.props);
  }

  componentWillReceiveProps(nextProps, nextContext) {
    const { updatedStipend : { isSaving }} = nextProps;
    this.setState({errors: nextProps.editing
      ? nextProps.travelStipends.updatedStipend.errors
      : nextProps.travelStipends.error });
    if (!isSaving) {
      this.updateStateData(nextProps);
    }
  }

  updateStateData = ({ editing, travelStipends: { selectedStipend } }) => {
    if(editing && selectedStipend ){
      const { amount:  stipend, center: { location } } = selectedStipend;
      this.setState({
        values: {
          center: location,
          stipend
        }
      });
    }
  }

  handleSubmit = event => {
    event.preventDefault();
    let { values, hasBlankFields } = this.state;
    if (!hasBlankFields) {
      const {
        handleCreateTravelStipend,
        history,
        updateTravelStipend,
        editing,
        travelStipends: { selectedStipend: { id } }
      } = this.props;

      if (editing) {
        updateTravelStipend(id, values, history);
      } else {
        handleCreateTravelStipend(values, history);
      }
    }
  };

  handleShowEventError = (event) => {
    event.preventDefault();
    this.setState({ isValidAmount: false });
  }

  handleOnchange = (event) => {
    const { target: { value: amount } } = event;  
    this.setState(prevSate => {
      return {
        values: {
          ...prevSate.values,
          stipend: amount
        },
        errors: {},
        isValidAmount: (amount ? (amount > 0 && amount <= 1000)  : true),
      };
    }, () => this.validate());
  }

  renderTravelStipendFieldset = (isEmpty) => {
    const { values, isValidAmount,  } = this.state;
    const { centers, editing, travelStipends } = this.props;
    return (
      <TravelStipendFieldset
        centers={centers}
        values={values}
        isEmpty={isEmpty}
        handleShowEventError={this.handleShowEventError}
        onChangeAmountInput={this.handleOnchange}
        isValidAmount={isValidAmount}
        value="245px"
        editing={editing}
        stipends={travelStipends.stipends}
      />
    );
  };

  renderForm = () => {
    const { errors, values, 
      hasBlankFields, 
      selection, 
      errors: { 
        stipend 
      }, 
      isValidAmount 
    } = this.state;
    const {
      closeModal,
      travelStipends,
      editing,
      updatedStipend: { isSaving }
    } = this.props;
    const isEmpty = stipend === 'This field is required';
    return (
      <FormContext
        targetForm={this}
        values={values}
        errors={errors}
        validatorName="validate">
        <form onSubmit={this.handleSubmit} className="new-request">
          {this.renderTravelStipendFieldset(isEmpty)}
          <SubmitArea
            onCancel={closeModal}
            travelStipends={travelStipends} 
            hasBlankFields={hasBlankFields || !isValidAmount}
            selection={selection}
            loading={isSaving}
            send={editing ? 'Save Stipend' : 'Add Stipend'} />
        </form>
      </FormContext>
    );
  };

  render() {
    return <div>{this.renderForm()}</div>;
  }
}

NewTravelStipendForm.propTypes = {
  closeModal: PropTypes.func,
  centers: PropTypes.array,
  travelStipends: PropTypes.object,
  handleCreateTravelStipend: PropTypes.func.isRequired,
  history: PropTypes.object,
  editing: PropTypes.bool,
  updateTravelStipend: PropTypes.func,
  updatedStipend: PropTypes.object
};

NewTravelStipendForm.defaultProps = {
  centers: [],
  closeModal: null,
  travelStipends: {},
  history: {
    push : () => {}
  },
  editing: false,
  updateTravelStipend: () => {},
  updatedStipend: {
    isSaving: false
  }
};

export default NewTravelStipendForm;
