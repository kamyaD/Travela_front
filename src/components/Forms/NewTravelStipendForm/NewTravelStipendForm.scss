form.new-request {
  .stipend-amount {
    input {
      width: 421px;
    }

    margin-right: 10px;
  }

  .stipend-location {
    width: 421px;

    .select-dropdown {
      margin-right: 10px;
    }
  }

}