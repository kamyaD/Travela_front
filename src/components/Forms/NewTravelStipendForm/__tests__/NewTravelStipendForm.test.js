import React from 'react';
import NewTravelStipendForm from '../NewTravelStipendForm';
import profileMock from '../../ProfileForm/__mocks__';


const { centers } = profileMock;

describe('<NewTravelStipendForm />', () => {
  let wrapper, onSubmit;
  onSubmit = jest.fn();

  const props = {
    loading: false,
    travelStipends: { isLoading: false },
    hasBlankFields: false,
    user: {
      UserInfo: {
        id: '-LJNw1BsT0LP_E4l2peP',
        name: 'Collins',
      }
    },
    userData:{
      id: '2',
      fullName: 'Collins Muru',
      name: 'Collins',
      email: 'collins.muru@andela.com',
      userId: '-LJNw1BsT0LP_E4l2peP',
      passportName: 'Collins',
      department: 'Talent & Development',
      occupation: 'Software Developer',
      manager: 'Collins',
      gender: 'Male',
      createdAt: '2018-09-14T12:48:11.266Z',
      updatedAt: '2018-09-16T07:53:48.835Z',
      roleId: 401938
    },
    userDataUpdate:[],
    handleCreateTravelStipend: jest.fn(() => {}),
    closeModal: jest.fn(),
    choices: ['kigali', 'nairobi'],
    modalType:'create travel stipend',
    centers
  };


  beforeEach(() => {
    wrapper = mount(<NewTravelStipendForm {...props} />);
    jest.resetAllMocks();
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders two fieldsets', () => {
    const fieldsets = wrapper.find('fieldset');
    expect(fieldsets).toHaveLength(2);
  });

  it('renders the travel stipend fieldset', () => {
    const travelStipendDetails = wrapper.find('fieldset .submit-area');
    expect(travelStipendDetails).toHaveLength(1);
  });

  it('renders two buttons', () => {
    const buttons = wrapper.find('button');
    expect(buttons).toHaveLength(2);
  });

  it('checks that onSubmit is not called when form is empty', () => {
    const form = wrapper.find('form');
    form.simulate('submit');
    expect(onSubmit).toHaveBeenCalledTimes(0);
  });
});

