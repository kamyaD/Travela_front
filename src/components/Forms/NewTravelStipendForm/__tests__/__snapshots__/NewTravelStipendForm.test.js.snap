// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewTravelStipendForm /> renders correctly 1`] = `
<div>
  <form
    className="new-request"
    onSubmit={[Function]}
  >
    <fieldset
      className="personal-details"
    >
      <div>
        <div>
          <div
            className="input-group"
          >
            <div
              className="spaces"
            >
              <div
                className="form-input request_dropdown stipend-location "
              >
                <label
                  htmlFor="center"
                >
                  Select Location
                  <span
                    style={
                      Object {
                        "color": "red",
                      }
                    }
                  >
                     * 
                  </span>
                </label>
                <div
                  style={
                    Object {
                      "position": "relative",
                    }
                  }
                >
                  <div
                    autoComplete="off"
                    choices={
                      Array [
                        "Lagos, Nigeria",
                        "Nairobi, Kenya",
                        "Kigali, Rwanda",
                        "New York, United States",
                        "Kampala, Uganda",
                        "San Francisco, United States",
                        "Austin, United States",
                      ]
                    }
                    className="input select  request_dropdown stipend-location"
                    id="user-location"
                    label="Select Location"
                    name="center"
                    onChange={[Function]}
                    onClick={[Function]}
                    onKeyPress={[Function]}
                    role="button"
                    size=""
                    tabIndex="0"
                    type="dropdown-select"
                    value=""
                  >
                    <div
                      className="value"
                      style={
                        Object {
                          "width": "",
                        }
                      }
                    >
                       
                      <div
                        className="select-dropdown inactive"
                      >
                        <img
                          alt="icn"
                          src="form_select_dropdown.svg"
                        />
                      </div>
                    </div>
                    <ul
                      className="select-menu select-menu--inactive"
                      onScroll={[Function]}
                    >
                      <li
                        id="choice"
                        key="Lagos, Nigeria"
                      >
                        <div
                          id="choice"
                          onClick={[Function]}
                          onKeyPress={[Function]}
                          role="button"
                          tabIndex="0"
                        >
                          Lagos, Nigeria
                        </div>
                      </li>
                      <li
                        id="choice"
                        key="Nairobi, Kenya"
                      >
                        <div
                          id="choice"
                          onClick={[Function]}
                          onKeyPress={[Function]}
                          role="button"
                          tabIndex="0"
                        >
                          Nairobi, Kenya
                        </div>
                      </li>
                      <li
                        id="choice"
                        key="Kigali, Rwanda"
                      >
                        <div
                          id="choice"
                          onClick={[Function]}
                          onKeyPress={[Function]}
                          role="button"
                          tabIndex="0"
                        >
                          Kigali, Rwanda
                        </div>
                      </li>
                      <li
                        id="choice"
                        key="New York, United States"
                      >
                        <div
                          id="choice"
                          onClick={[Function]}
                          onKeyPress={[Function]}
                          role="button"
                          tabIndex="0"
                        >
                          New York, United States
                        </div>
                      </li>
                      <li
                        id="choice"
                        key="Kampala, Uganda"
                      >
                        <div
                          id="choice"
                          onClick={[Function]}
                          onKeyPress={[Function]}
                          role="button"
                          tabIndex="0"
                        >
                          Kampala, Uganda
                        </div>
                      </li>
                      <li
                        id="choice"
                        key="San Francisco, United States"
                      >
                        <div
                          id="choice"
                          onClick={[Function]}
                          onKeyPress={[Function]}
                          role="button"
                          tabIndex="0"
                        >
                          San Francisco, United States
                        </div>
                      </li>
                      <li
                        id="choice"
                        key="Austin, United States"
                      >
                        <div
                          id="choice"
                          onClick={[Function]}
                          onKeyPress={[Function]}
                          role="button"
                          tabIndex="0"
                        >
                          Austin, United States
                        </div>
                      </li>
                    </ul>
                  </div>
                </div>
                <span
                  className="error"
                />
              </div>
            </div>
            <div
              className="spaces"
            >
              <div
                className="form-input request_dropdown stipend-amount "
              >
                <label
                  htmlFor="stipend"
                >
                  Enter Amount in Dollars ($)
                  <span
                    style={
                      Object {
                        "color": "red",
                      }
                    }
                  >
                     * 
                  </span>
                </label>
                <input
                  autoComplete="off"
                  className=" request_dropdown stipend-amount"
                  id="your-manager"
                  label="Enter Amount in Dollars ($)"
                  name="stipend"
                  onBlur={[Function]}
                  onChange={[Function]}
                  size=""
                  type="text"
                  value=""
                />
                <span
                  className="error"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </fieldset>
    <fieldset>
      <div
        className="submit-area"
      >
        <button
          className="bg-btn bg-btn--inactive travel-stipends__right"
          id="cancel"
          onClick={[Function]}
          type="button"
        >
          Cancel
        </button>
        <button
          className="bg-btn bg-btn--active"
          disabled={true}
          id="submit"
          type="submit"
        >
          <undefined />
        </button>
      </div>
    </fieldset>
  </form>
</div>
`;
